main/src/ca/uwaterloo/flix/language/phase/optimizer/Inliner.scala

```diff
@@ -371,6 +373,195 @@ object Inliner {
+        MatchResult.Match(binders1 ++ binders2)
+    }
+
+  }
+
+  /** Returns the match result of `exp` against `pat` with the guard `guardOpt`. */
+  private def matchRule(exp: MonoAst.Expr, pat: Pattern, guardOpt: Option[MonoAst.Expr]): MatchResult = {
+    val guardVal = guardOpt.map(evalBoolExpression).getOrElse(FuzzyBool.True)
+    (matchPat(exp, pat), guardVal) match {
+      case (_, FuzzyBool.False) | (MatchResult.NoMatch, _) =>
+        MatchResult.NoMatch
+      case (MatchResult.Unknown, _) | (MatchResult.Match(_), FuzzyBool.Unknown) =>
+        MatchResult.Unknown
+      case (resultMatch@MatchResult.Match(_), FuzzyBool.True) =>
+        resultMatch
+    }
+  }
+
+  /** Returns the match result of `exp` against `pat`. */
+  private def matchPat(exp: MonoAst.Expr, pat: MonoAst.Pattern): MatchResult = pat match {
+    case Pattern.Wild(_, _) =>
+      // Preserve the expression in case it is impure.
+      MatchResult.singleMatch(None, exp)
+
+    case v@Pattern.Var(_, _, _, _) =>
+      MatchResult.singleMatch(Some(v), exp)
+
+    case Pattern.Cst(cst, _, _) =>
+      exp match {
+        case Expr.Cst(expCst, _, _) =>
+          matchConstant(cst, expCst)
+        case _ =>
+          MatchResult.Unknown
+      }
+
+    case Pattern.Tag(symUse, pats, _, _) =>
+      exp match {
+        case Expr.ApplyAtomic(AtomicOp.Tag(caseSym), exps, _, _, _) =>
+          if (symUse.sym != caseSym) MatchResult.NoMatch
+          else matchPatterns(exps, pats)
+        case _ =>
+          MatchResult.Unknown
+      }
+
+    case Pattern.Tuple(pats, _, _) =>
+      exp match {
+        case Expr.ApplyAtomic(AtomicOp.Tuple, exps, _, _, _) =>
+          matchPatterns(exps, pats.toList)
+        case _ =>
+          MatchResult.Unknown
+      }
+
+    case Pattern.Record(_, _, _, _) =>
+      MatchResult.Unknown
+
+  }
+
+  /**
+    * Returns the match result for matching `exps` to the patterns of `pats`.
+    *
+    * If the two lists are of different lengths, [[MatchResult.Unknown]] is returned.
+    */
+  private def matchPatterns(exps: List[MonoAst.Expr], pats: List[MonoAst.Pattern]): MatchResult = {
+    if (exps.lengthCompare(pats) != 0) {
+      MatchResult.Unknown
+    } else {
+      val res = pats.zip(exps).foldLeft(MatchResult.emptyMatch()) {
+        case (acc, (innerPat, innerExp)) =>
+          MatchResult.concat(acc, matchPat(innerExp, innerPat))
+      }
+      res
+    }
+  }
+
+  /**
+    * Returns the match result for two constants.
+    *
+    * Constants contain no binders so the function is symmetric.
+    */
+  private def matchConstant(cst1: Constant, cst2: Constant): MatchResult = {
+    import Constant.*
+    import MatchResult.*
+    (cst1, cst2) match {
+      case (Unit, Unit) => emptyMatch()
+      case (Null, Null) => emptyMatch()
+      case (Bool(v1), Bool(v2)) => matchFromBool(v1 == v2)
+      case (Char(v1), Char(v2)) => matchFromBool(v1 == v2)
+      case (Float32(v1), Float32(v2)) => matchFromBool(v1 == v2)
```

Comment:
Thoughts on your best friend, NaN?

Reply:
`v1 == v2` is using float equality, just like a pattern match would do. So no matter what happens it should be consistent.